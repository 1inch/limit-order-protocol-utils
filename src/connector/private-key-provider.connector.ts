import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';
import {MsgParams, signTypedData_v4} from 'eth-sig-util';

export class PrivateKeyProviderConnector implements ProviderConnector {
    constructor(
        private readonly privateKey: string,
        protected readonly web3Provider: Web3
    ) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        _walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash = ''
    ): Promise<string> {
        const result = signTypedData_v4(Buffer.from(this.privateKey, 'hex'), {
            data: typedData,
            /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        } as MsgParams<any>);

        return Promise.resolve(result);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
