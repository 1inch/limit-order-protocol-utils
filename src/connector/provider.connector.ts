import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem, AbiOutput} from '../model/abi.model';

export interface AbiDecodeResult {
    [key: string]: unknown,
}

export interface ProviderConnector {
    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string;

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        typedDataHash: string
    ): Promise<string>;

    ethCall(contractAddress: string, callData: string): Promise<string>;

    decodeABIParameter<T>(type: AbiOutput | string, hex: string): T;

    decodeABICallParameters(types: Array<AbiOutput | string>, callData: string): AbiDecodeResult;
}
