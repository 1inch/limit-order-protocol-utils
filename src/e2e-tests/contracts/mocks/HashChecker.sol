// SPDX-License-Identifier: MIT

pragma solidity 0.8.19;

import "@openzeppelin/contracts/access/Ownable.sol";
import "../interfaces/IPreInteraction.sol";
import "../OrderLib.sol";


contract HashChecker is IPreInteraction, Ownable {
    using OrderLib for IOrderMixin.Order;

    error IncorrectOrderHash();

    bytes32 public immutable limitOrderProtocolDomainSeparator;
    mapping(bytes32 => bool) public hashes;

    constructor (address limitOrderProtocol) {
        // solhint-disable-next-line avoid-low-level-calls
        (, bytes memory data) = limitOrderProtocol.call(abi.encodeWithSignature("DOMAIN_SEPARATOR()"));
        limitOrderProtocolDomainSeparator = abi.decode(data, (bytes32));
    }

    function setHashOrderStatus(IOrderMixin.Order calldata order, bool status) external onlyOwner {
        bytes32 orderHash = order.hash(limitOrderProtocolDomainSeparator);
        hashes[orderHash] = status;
    }

    function preInteraction(
        IOrderMixin.Order calldata order,
        bytes32 orderHash,
        address taker,
        uint256 makingAmount,
        uint256 takingAmount,
        uint256 remainingMakingAmount,
        bytes calldata extraData
    ) external override {
        if (hashes[orderHash] == false) revert IncorrectOrderHash();

        if (extraData.length != 0) {
            IPreInteraction(address(bytes20(extraData))).preInteraction(
                order,
                orderHash,
                taker,
                makingAmount,
                takingAmount,
                remainingMakingAmount,
                extraData[20:]
            );
        }
    }
}
