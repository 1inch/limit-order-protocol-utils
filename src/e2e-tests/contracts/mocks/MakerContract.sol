// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

import "@openzeppelin/contracts/interfaces/IERC1271.sol";
import "@openzeppelin/contracts/token/ERC20/ERC20.sol";

import "../LimitOrderProtocol.sol";
import "../OrderLib.sol";
import { EIP712Alien } from "./EIP712Alien.sol";

contract MakerContract is IERC1271, EIP712Alien, ERC20 {
    using SafeERC20 for IERC20;
    using OrderLib for IOrderMixin.Order;
    using AddressLib for Address;
    using MakerTraitsLib for MakerTraits;

    error NotAllowedToken();
    error BadPrice();
    error MalformedSignature();

    address immutable public protocol;
    IERC20 immutable public token0;
    IERC20 immutable public token1;
    uint256 immutable public fee;
    uint256 immutable public fee2;

    constructor(
        address _protocol,
        IERC20 _token0,
        IERC20 _token1,
        uint256 _fee,
        string memory name,
        string memory symbol
    )
        EIP712Alien(_protocol, "1inch Limit Order Protocol", "4")
        ERC20(name, symbol)
    {
        protocol = _protocol;
        token0 = _token0;
        token1 = _token1;
        fee = _fee;
        fee2 = 2e18 * _fee / (1e18 + _fee);
        _token0.approve(_protocol, type(uint256).max);
        _token1.approve(_protocol, type(uint256).max);
    }

    function decimals() public pure override returns (uint8) {
        return 6;
    }

    function deposit(IERC20 token, uint256 amount) public {
        depositFor(token, amount, msg.sender);
    }

    function depositFor(IERC20 token, uint256 amount, address to) public {
        if (token != token0 && token != token1) revert NotAllowedToken();

        _mint(to, amount * fee2 / 1e18);
        token.safeTransferFrom(msg.sender, address(this), amount);
    }

    function withdraw(IERC20 token, uint256 amount) public {
        withdrawFor(token, amount, msg.sender);
    }

    function withdrawFor(IERC20 token, uint256 amount, address to) public {
        if (token != token0 && token != token1) revert NotAllowedToken();

        _burn(msg.sender, amount);
        token.safeTransfer(to, amount * fee2 / 1e18);
    }

    function isValidSignature(bytes32 hash, bytes calldata signature) external view override returns(bytes4) {
        if (signature.length != 8 * 0x20) revert MalformedSignature();

        IOrderMixin.Order calldata order;
        assembly ("memory-safe") { // solhint-disable-line no-inline-assembly
            order := signature.offset
        }

        if (
            (
                (order.makerAsset.get() != address(token0) || order.takerAsset.get() != address(token1)) &&
                (order.makerAsset.get() != address(token1) || order.takerAsset.get() != address(token0))
            ) ||
            order.makerTraits.hasExtension() ||
            order.makingAmount * fee > order.takingAmount * 1e18 ||
            order.hash(_domainSeparatorV4()) != hash
        ) revert BadPrice();

        return this.isValidSignature.selector;
    }
}
