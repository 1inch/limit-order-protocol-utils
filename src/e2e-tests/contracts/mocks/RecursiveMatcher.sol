// SPDX-License-Identifier: MIT

pragma solidity 0.8.19;

import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import "../interfaces/ITakerInteraction.sol";
import "../interfaces/IOrderMixin.sol";
import "../interfaces/IOrderMixin.sol";
import "../libraries/TakerTraitsLib.sol";

contract RecursiveMatcher is ITakerInteraction {
    bytes1 private constant _FINALIZE_INTERACTION = 0x01;

    error IncorrectCalldataParams();
    error FailedExternalCall(bytes reason);

    function matchOrders(
        IOrderMixin orderMixin,
        IOrderMixin.Order calldata order,
        bytes32 r,
        bytes32 vs,
        uint256 amount,
        TakerTraits takerTraits,
        bytes calldata interaction
    ) external {
        orderMixin.fillOrderTo(
            order,
            r,
            vs,
            amount,
            takerTraits,
            address(this),
            interaction
        );
    }

    function takerInteraction(
        IOrderMixin.Order calldata /* order */,
        bytes32 /* orderHash */,
        bytes calldata /* extension */,
        address /* taker */,
        uint256 /* makingAmount */,
        uint256 /* takingAmount */,
        uint256 /* remainingMakingAmount */,
        bytes calldata extraData
    ) external returns(uint256 offeredTakingAmount) {
        if(extraData[0] & _FINALIZE_INTERACTION != 0x0) {
            (
                address[] memory targets,
                bytes[] memory calldatas
            ) = abi.decode(extraData[1:], (address[], bytes[]));

            if (targets.length != calldatas.length) revert IncorrectCalldataParams();
            for (uint256 i = 0; i < targets.length; i++) {
                // solhint-disable-next-line avoid-low-level-calls
                (bool success, bytes memory reason) = targets[i].call(calldatas[i]);
                if(!success) revert FailedExternalCall(reason);
            }
        } else {
            // solhint-disable-next-line avoid-low-level-calls
            (bool success, bytes memory reason) = msg.sender.call(
                abi.encodePacked(IOrderMixin.fillOrderTo.selector, extraData[1:])
            );
            if (!success) revert FailedExternalCall(reason);
        }
        return 0;
    }
}
