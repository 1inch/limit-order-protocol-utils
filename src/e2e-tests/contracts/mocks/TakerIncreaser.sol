// SPDX-License-Identifier: MIT

pragma solidity 0.8.19;

import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import "../interfaces/ITakerInteraction.sol";
import "../interfaces/IOrderMixin.sol";

contract TakerIncreaser is ITakerInteraction {
    error IncorrectCalldataParams();
    error FailedExternalCall();

    function fillOrderTo(
        IOrderMixin orderMixin,
        IOrderMixin.Order calldata order,
        bytes32 r,
        bytes32 vs,
        uint256 amount,
        TakerTraits takerTraits,
        address target,
        bytes calldata interaction
    ) external {
        orderMixin.fillOrderTo(
            order,
            r,
            vs,
            amount,
            takerTraits,
            target,
            interaction
        );
    }

    function takerInteraction(
        IOrderMixin.Order calldata /* order */,
        bytes32 /* orderHash */,
        bytes calldata /* extension */,
        address /* taker */,
        uint256 /* makingAmount */,
        uint256 takingAmount,
        uint256 /* remainingMakingAmount */,
        bytes calldata extraData
    ) external returns(uint256) {
        if (extraData.length != 0) {
            (
                address[] memory targets,
                bytes[] memory calldatas
            ) = abi.decode(extraData, (address[], bytes[]));
            if (targets.length != calldatas.length) revert IncorrectCalldataParams();
            for (uint256 i = 0; i < targets.length; i++) {
                // solhint-disable-next-line avoid-low-level-calls
                (bool success, ) = targets[i].call(calldatas[i]);
                if(!success) revert FailedExternalCall();
            }
        }
        return takingAmount * 15 / 10; // increase 50% taking amount
    }
}
