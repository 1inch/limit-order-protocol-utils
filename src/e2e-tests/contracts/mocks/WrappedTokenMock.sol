// SPDX-License-Identifier: MIT

pragma solidity 0.8.19;

import "@openzeppelin/contracts/access/Ownable.sol";
import "@openzeppelin/contracts/token/ERC20/extensions/draft-ERC20Permit.sol";
import "@1inch/solidity-utils/contracts/interfaces/IWETH.sol";

/// @title Generic token for testing purposes with deposit/withdraw capabilities
contract WrappedTokenMock is ERC20Permit, Ownable, IWETH {
    error NotEnoughBalance();

    event Deposit(address indexed dst, uint wad);
    event Withdrawal(address indexed src, uint wad);

    // solhint-disable-next-line no-empty-blocks
    constructor(string memory name, string memory symbol) ERC20(name, symbol) ERC20Permit(name) {}

    // solhint-disable-next-line no-empty-blocks
    receive() external payable {
        deposit();
    }

    function getChainId() external view returns (uint256) {
        return block.chainid;
    }

    function deposit() public payable {
        _mint(msg.sender, msg.value);
        // balanceOf[msg.sender] += msg.value;
        emit Deposit(msg.sender, msg.value);
    }

    function withdraw(uint wad) public {
        if (balanceOf(msg.sender) < wad) revert NotEnoughBalance();
        _burn(msg.sender, wad);
        payable(msg.sender).transfer(wad);
        emit Withdrawal(msg.sender, wad);
    }
}
