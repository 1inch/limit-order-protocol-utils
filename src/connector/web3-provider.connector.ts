import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';

export class Web3ProviderConnector implements ProviderConnector {
    constructor(protected readonly web3Provider: Web3) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash: string
    ): Promise<string> {
        if (!this.web3Provider.currentProvider) {
            throw new Error('Web3 currentProvider is null');
        }

        return this.web3Provider.currentProvider.send('eth_signTypedData_v4', [
            walletAddress,
            JSON.stringify(typedData),
        ]);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
