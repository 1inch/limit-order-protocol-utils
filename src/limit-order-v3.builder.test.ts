import {LimitOrderDecoder} from './limit-order.decoder';
import {LimitOrderLegacy} from './model/limit-order-protocol.model';
import Web3 from 'web3';
import {PrivateKeyProviderConnector} from './connector/private-key-provider.connector';
import {limitOrderProtocolAddresses} from './limit-order-protocol.const';
import {largeInteractions, largeResult} from './test/mocks';
import {LimitOrderV3Builder} from "./limit-order-v3.builder";

describe('LimitOrderV3Builder - for build new limit order', () => {
    const chainId = 56;
    const contractAddress = limitOrderProtocolAddresses[chainId];

    const web3 = new Web3('https://bsc-dataseed.binance.org');
    const privateKey =
        '552be66668d14242eeeb0e84600f0946ddddc77777777c3761ea5906e9ddcccc';
    const providerConnector = new PrivateKeyProviderConnector(privateKey, web3);

    let limitOrderBuilder: LimitOrderV3Builder;

    beforeEach(() => {
        limitOrderBuilder = new LimitOrderV3Builder(
            providerConnector,
            {
                domainName: '1inch Aggregation Router',
                version: '5'
            }
        );
    });

    describe('Normal limit order', () => {
        it('buildOrderSignature() must call the provider signTypedData method', async () => {
            const walletAddress = '0x1548dAdf412Eaaf3c80bEad35CDa83a4bf7dF6ce';
            const dataHash =
                '7cb7e268d5a5f0d8da9a5904a0084b3c4f17a7826413e83d69784a50d4154878';

            const { interactions, offsets } = LimitOrderV3Builder.packInteractions({
                makerAssetData: '0xf0',
                takerAssetData: '0xf1',
                getMakingAmount: '0xf2',
                getTakingAmount: '0xf3',
                predicate: '0xf4',
                permit: '0xf5',
                preInteraction: '0xf6',
                postInteraction: '0xf7',
            });

            const order: LimitOrderLegacy = {
                salt: '1',
                makerAsset: 'makerAsset',
                takerAsset: 'takerAsset',
                maker: 'maker',
                receiver: 'receiver',
                allowedSender: 'allowedSender',
                makingAmount: 'makingAmount',
                takingAmount: 'takingAmount',

                offsets,
                interactions,
            };
            const typedData = limitOrderBuilder.buildLimitOrderTypedData(
                order,
                chainId,
                contractAddress
            );

            const signTypedDataSpy = jest.spyOn(
                providerConnector,
                'signTypedData'
            ).and.callThrough();

            const signature = await limitOrderBuilder.buildOrderSignature(
                walletAddress,
                typedData
            );

            expect(signature).toMatchSnapshot();
            expect(signTypedDataSpy).toHaveBeenCalledTimes(1);
            expect(signTypedDataSpy).toHaveBeenCalledWith(
                walletAddress,
                typedData,
                dataHash,
            );
        });

        it('buildLimitOrderHash() must create a hash of order with 0x prefix', () => {
            const { interactions, offsets } = LimitOrderV3Builder.packInteractions({
                makerAssetData: '0xf0',
                takerAssetData: '0xf1',
                getMakingAmount: '0xf2',
                getTakingAmount: '0xf3',
                predicate: '0xf4',
                permit: '0xf5',
                preInteraction: '0xf6',
                postInteraction: '0xf7',
            });

            const order: LimitOrderLegacy = {
                salt: '1',
                makerAsset: 'makerAsset',
                takerAsset: 'takerAsset',
                maker: 'maker',
                receiver: 'receiver',
                allowedSender: 'allowedSender',
                makingAmount: 'makingAmount',
                takingAmount: 'takingAmount',

                offsets,
                interactions,
            };
            const typedData = limitOrderBuilder.buildLimitOrderTypedData(
                order,
                chainId,
                contractAddress
            );

            const hash = limitOrderBuilder.buildLimitOrderHash(typedData);

            expect(hash).toMatchSnapshot();
        });

        it('buildLimitOrder() must create a limit order instance according to the given parameters', async () => {
            const makerAddress = '0xdddd91605c18a9999c1d47abfeed5daaaa700000';

            const order = limitOrderBuilder.buildLimitOrder({
                makerAssetAddress: '0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c',
                takerAssetAddress: '0x111111111117dc0aa78b770fa6a738034120c302',
                makerAddress,
                makingAmount: '3',
                takingAmount: '100',
                predicate: '0x11',
                permit: '0x22',
                postInteraction: '0x33',
            });

            expect(
                LimitOrderDecoder.unpackInteractionsV3(order.offsets, order.interactions)
            ).toMatchSnapshot();

            expect(+order.salt).toBeGreaterThan(1);

            order.salt = '1';

            expect(order).toMatchSnapshot();
        });
    });

    describe("packInteractions", () => {
            it("should pack", () => {
                const { offsets, interactions } = LimitOrderV3Builder.packInteractions(largeInteractions);

                expect(offsets).toBe(largeResult.offsets);
                expect(interactions).toBe(largeResult.interactions);
            })
    })
});
