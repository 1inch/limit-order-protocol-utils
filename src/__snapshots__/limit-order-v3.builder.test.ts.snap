// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LimitOrderV3Builder - for build new limit order Normal limit order buildLimitOrder() must create a limit order instance according to the given parameters 1`] = `
Object {
  "getMakingAmount": "0x",
  "getTakingAmount": "0x",
  "makerAssetData": "0x",
  "permit": "0x22",
  "postInteraction": "0x33",
  "preInteraction": "0x",
  "predicate": "0x11",
  "takerAssetData": "0x",
}
`;

exports[`LimitOrderV3Builder - for build new limit order Normal limit order buildLimitOrder() must create a limit order instance according to the given parameters 2`] = `
Object {
  "allowedSender": "0x0000000000000000000000000000000000000000",
  "interactions": "0x112233",
  "maker": "0xdddd91605c18a9999c1d47abfeed5daaaa700000",
  "makerAsset": "0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c",
  "makingAmount": "3",
  "offsets": "0x300000002000000020000000100000000000000000000000000000000",
  "receiver": "0x0000000000000000000000000000000000000000",
  "salt": "1",
  "takerAsset": "0x111111111117dc0aa78b770fa6a738034120c302",
  "takingAmount": "100",
}
`;

exports[`LimitOrderV3Builder - for build new limit order Normal limit order buildLimitOrderHash() must create a hash of order with 0x prefix 1`] = `"0xb817fcd255f7253abdc56c0686d42eb49fe2fd91efa5148953d23dfa46f9fd81"`;

exports[`LimitOrderV3Builder - for build new limit order Normal limit order buildOrderSignature() must call the provider signTypedData method 1`] = `"0x05712732e0cd9dfb4fdfcda3be3f62a734c1169474c5824816420d313cf6b33b35a91cf113ba11693793785914e408543b4c9418a8d10fc467b2709ed38c011a1b"`;
